File: backend-csharp/modules/payments/Internal/PaymentService.cs
// Business logic — hidden from the outside world.
using System.Collections.Concurrent;
namespace BackendCsharp.Modules.Payments.Internal;

public class Payment
{
    public string PaymentId { get; set; } = "";
    public string OrderId { get; set; } = "";
    public string Status { get; set; } = "";
    public long AmountCents { get; set; }
}

public class PaymentService
{
    private static readonly ConcurrentDictionary<string, Payment> Store = new();

    public Payment Authorize(string paymentId, string orderId, long amountCents)
    {
        var p = new Payment { PaymentId = paymentId, OrderId = orderId, Status = "AUTHORIZED", AmountCents = amountCents };
        Store[paymentId] = p;
        return p;
    }

    public Payment? Capture(string paymentId)
    {
        if (!Store.TryGetValue(paymentId, out var p) || p.Status != "AUTHORIZED") return null;
        p.Status = "CAPTURED";
        return p;
    }

    public Payment? Refund(string paymentId)
    {
        if (!Store.TryGetValue(paymentId, out var p) || p.Status != "CAPTURED") return null;
        p.Status = "REFUNDED";
        return p;
    }
}

File: backend-ocaml/modules/payments/internal/payment_service.ml
(* Business logic — hidden from the outside world. *)
type payment = { payment_id : string; order_id : string; mutable status : string; amount_cents : int }

let store : (string, payment) Hashtbl.t = Hashtbl.create 16

let authorize payment_id order_id amount_cents =
  let p = { payment_id; order_id; status = "AUTHORIZED"; amount_cents } in
  Hashtbl.replace store payment_id p;
  p

let capture payment_id =
  match Hashtbl.find_opt store payment_id with
  | Some p when p.status = "AUTHORIZED" -> p.status <- "CAPTURED"; Some p
  | _ -> None

let refund payment_id =
  match Hashtbl.find_opt store payment_id with
  | Some p when p.status = "CAPTURED" -> p.status <- "REFUNDED"; Some p
  | _ -> None
